### ISSUE TITLE
License Motivation for Open Source Projects

### ISSUE THREAD

**Issue by mewmew** [keystone-engine/keystone#161]
I would like to take this opportunity to ask the authors of keystone for their rationale and the motivation behind the chosen license for open source use (i.e. GPLv2).

Specifically, I'd be interested to know if other licenses were considered? As the indiegogo campaign mentioned that a permissive license would be used for open source projects, and since the LLVM project uses a BSD license, I had expected keystone to use a BSD-style license (e.g. BSD 1-, 2- or 3-clause, or MIT).

Please note, I am interested in the motivation behind the license for open source projects. I do not have any concerns with the dual license model, and would therefore like to focus the discussions in this topic to the license for open source projects.

Kind regards /u

Extract from the [indiegogo campaign](https://www.indiegogo.com/projects/keystone-open-source-assembler-framework--2):
> Distributed with full source code under dual license - with permissive license for open source projects.

**Comment by aquynh** [comment]
yes, we keep the Indiegogo promise, as the license of Keystone is very permissive.

this is from our README.md

> Version 2 of the GNU General Public License (GPLv2). (I.e. Without the "any later version" clause.). License information can be found in the COPYING file and the EXCEPTIONS-CLIENT file.

> This combination allows almost all of open source projects to use Keystone without conflicts.

the **exception** here means if your tool uses **any open source license**, Keystone license is (most likely) compatible with your tool (even though Keystone license is GPLv2). so you can use Keystone in your tool without any issues. see https://github.com/keystone-engine/keystone/blob/master/EXCEPTIONS-CLIENT for more information.

note that we do not change the license of the LLVM code: everything under `llvm/` directory (except `llvm/keystone/`) is still under the same LLVM (BSD) license.

regarding the motivation behind the license of Keystone, i think it is clearly addressed in our Indiegogo page. 

> We choose this dual-license for Keystone because from our experiences, we can tell that without any resource, it is very tough to maintain the quality of an open source project, even if it is initially successful. The dual license is the only way for us to get some fund, even minimal, to keep working on Keystone and push it to another level so everybody benefits from that.

hope this answers your question, thanks!

**Comment by mewmew** [comment]
> note that we do not change the license of the LLVM code: everything under llvm/ directory (except llvm/keystone/) is still under the same LLVM (BSD) license.

> regarding the motivation behind the license of Keystone, i think it is clearly addressed in our Indiegogo page.

> > We choose this dual-license for Keystone because from our experiences, we can tell that without any resource, it is very tough to maintain the quality of an open source project, even if it is initially successful. The dual license is the only way for us to get some fund, even minimal, to keep working on Keystone and push it to another level so everybody benefits from that.

> so Keystone is free for all open source projects. you only have to buy the license if your tool is commercial. i think this is fair & acceptable for everybody involved.

> hope this answers your question, thanks!

Thank your for the answer. Of course I understand the motivation of using a dual license, as it may help allocate time and resources for working on this project, which in and of itself will be of huge benefit to the open source community.

Allow me to provide some background on why I'm vary of considering GPL licensed software permissive. One requirement of GPL is that if binary copies of the software are hosted, then a copy of the source code has to be hosted as well. This turned out to cause quite a headache for rolling release open source distributions, such as Arch Linux; as described in the issue [ FS#5355 - Source code availability for full GPL compliance](https://webcache.googleusercontent.com/search?q=cache:j9WiZeBcB14J:https://bugs.archlinux.org/task/5355).

Therefore, let me to rephrase the question as follows, would it be possible to use an even more permissive license (i.e. BSD or MIT) while keeping the dual license model? Has this been done by anyone in the past? What were the results?

I'm very happy to see you working on this project, and I think it will have the potential of making a huge impact as its design is thorough and well thought-out, and since there are really no other open source assembler/disassembler project with the same ambitious aims. My main wish is that it will reach as many open source developers and possible, and that it's progress and spread should be dependent only on technical merits and not bureaucratical ones.

Best wishes and happy hacking.

Cheers /u

**Comment by aquynh** [comment]
that is a very good question. we just use the similar license with MySQL, so think it can work. other than that, i never seen any dual-license project with BSD as one license. i am not a lawyer, but i think that is not possible to use BSD or MIT in this case.

actually i would love to go for BSD if that can be the answer.

as for your concern about spreading Capstone & Keystone usage, i dont see any reason stopping open source projects from using them because both licenses are very permissive, as much as possible from what i can see.

**Comment by aquynh** [comment]
just to share a thought: initially i wanted to go for dual-license with (pure) GPL, but then i realized that it would prohibit many open source projects that use incompatible licenses - note that Keystone is a library. finally i chose the current model which is GPL + exception, so Keystone is as permissive as possible, and everybody is happy.

**Comment by mewmew** [comment]
While the post is opinionated, http://devblog.ailon.org/devblog/post/2011/12/27/GPLCommercial-Dual-Licensing-Is-Evil-For-Libraries.aspx provides some background from another project that tried to use a GPL/commercial dual license. May be an interesting read.

> actually i would love to go for BSD if that can be the answer.

I'm glad to hear! I don't see any reason why it couldn't? It would seem reasonable that the project could still make use a dual license model, no matter the license used for open source projects. Simply stating that a commercial license is required for commercial projects, while open source projects may use a BSD licensed version of Keystone. Then again, as I'm not a lawyer, this may be a naive perspective.

I feel happy that you have given it some serious thought, as a detail such as this can have far reaching consequences for the project.

As always, happy hacking :) /u

**Comment by aquynh** [comment]
i know very well that GPL is not a good license for a library, but with those *exception*, there is no more issues, isnt it?

**Comment by mewmew** [comment]
> i know very well that GPL is not a good license for a library, but with those *exception*, there is no more issues, isnt it?

I understand that the exception is an attempt at finding a sweet spot. However, I know many projects have very clear directives one the use of GPL/AGPL in external libraries, and one would have to ask each single one of these projects about their thoughts regarding the *exception*. Some may agree that it is enough, and allow the library to be used, while others may disagree, thus making it impossible to use Keystone for these projects.

Take NuPIC (Numenta Platform for Intelligent Computing) for instance, a machine learning and research framework with some incredibly interesting ideas. [From the NuPIC FAQ](http://numenta.org/faq/#can-i-add-an-external-library-to-nupic):

> Can I add an external library to NuPIC?
>
> You can add external libraries to NuPIC. They must have a license that permits proprietary use of the library. External libraries licensed under AGPL, for instance, will not be accepted.

The NuPIC framework is dealing with many different input sources, ranging from visual data, auditory, spatial and temporal data to continuous and discrete numbers. I'd be interested in adding support for native understanding of machine instructions, preferably using Keystone. This is one of the reasons I really hope that there may exist another alternative than using GPL for Keystone, while still allowing a dual license model to be used.

Sincere regards /u

**Comment by aquynh** [comment]
what exactly is their definition of "proprietary use of the library"?

and what is wrong with AGPL for NuPIC?

**Comment by crowell** [comment]
agreed, BSD or LGPL would really be ideal for this. Without getting into a big licensing debate, LGPL is used in libraries over GPL for a reason. I'd love to see a BSD licensed version of keystone.

**Comment by boboben1** [comment]
This library is too restrictive for my upcoming project also. I'd love too see this as LGPL or MIT.

**Comment by aquynh** [comment]
the idea is that: if you make money, contribute back to project, so the developers can continue developing it. is this unreasonable? if it is not, then why?

**Comment by aquynh** [comment]
regarding the license, LGPL is incompatible with many licenses, while this is not.

**Comment by boboben1** [comment]
How is LGPL more restrictive than GPL? Isn't that the point of LGPL?

**Comment by boboben1** [comment]
I believe I can use the FOSS exception to include this in an open source project under different licensing if my understanding is correct.

**Comment by aquynh** [comment]
This is not really GPL, you should look at the exceptions. That makes this much more permissive than GPL, or even LGPL (except LGPL does not require license fee from commercial products)

Basically you can combine this with whatever open source license code you are using. You only need to pay license fee if you commercialize your products.

**Comment by boboben1** [comment]
I am just unsure how the license would affect usage of my library (LGPL) in free, but closed source projects.

**Comment by aquynh** [comment]
That is OK to me

**Comment by SBvn-dev** [comment]
> yes, we keep the Indiegogo promise, as the license of Keystone is very permissive.
> 
> this is from our README.md
> 
> > Version 2 of the GNU General Public License (GPLv2). (I.e. Without the "any later version" clause.). License information can be found in the COPYING file and the EXCEPTIONS-CLIENT file.
> > This combination allows almost all of open source projects to use Keystone without conflicts.
> 
> the **exception** here means if your tool uses **any open source license**, Keystone license is (most likely) compatible with your tool (even though Keystone license is GPLv2). so you can use Keystone in your tool without any issues. see https://github.com/keystone-engine/keystone/blob/master/EXCEPTIONS-CLIENT for more information.
> 
> note that we do not change the license of the LLVM code: everything under `llvm/` directory (except `llvm/keystone/`) is still under the same LLVM (BSD) license.
> 
> regarding the motivation behind the license of Keystone, i think it is clearly addressed in our Indiegogo page.
> 
> > We choose this dual-license for Keystone because from our experiences, we can tell that without any resource, it is very tough to maintain the quality of an open source project, even if it is initially successful. The dual license is the only way for us to get some fund, even minimal, to keep working on Keystone and push it to another level so everybody benefits from that.
> 
> so Keystone is free for all open source projects. you only have to buy the license if your tool is commercial. i think this is fair & acceptable for everybody involved.
> 
> hope this answers your question, thanks!

@aquynh Is it allowed to use keystone.dll in a GPL-3 project? or does it require the project to change its' license to GPL-2?
